File: apps/android/core/model/src/main/kotlin/com/sigmundgranaas/turbo/expressive/domain/LocationDescription.kt
package com.sigmundgranaas.turbo.expressive.domain

/** How a place name relates to the queried point. */
enum class PlaceQualifier { On, In, Near, At }

/**
 * A human description of a coordinate, resolved by reverse-geocoding: a primary
 * [title] (the nearest meaningful place name), an optional [qualifier] ("On
 * Galdhøpiggen", "In Lom"), an optional [secondary] line (feature kind, kommune,
 * or address detail) and an optional [elevationM]. Mirrors the Flutter
 * `LocationDescription`.
 */
data class LocationDescription(
    val title: String,
    val qualifier: PlaceQualifier? = null,
    val secondary: String? = null,
    val elevationM: Double? = null,
) {
    /** Headline, e.g. "On Galdhøpiggen" / "In Lom". */
    val label: String
        get() = when (qualifier) {
            PlaceQualifier.On -> "On $title"
            PlaceQualifier.In -> "In $title"
            PlaceQualifier.Near -> "Near $title"
            PlaceQualifier.At, null -> title
        }

    /** Supporting line, e.g. "fjelltopp · 2469 m · Lom". */
    val subtitle: String
        get() = listOfNotNull(
            secondary?.takeIf(String::isNotBlank),
            elevationM?.let { "${it.toInt()} m" },
        ).joinToString(" · ")
}

File: apps/ios/Sources/CoreModel/LocationDescription.swift
import Foundation

/// How a place name relates to the queried point. Mirrors `domain.PlaceQualifier`.
public enum PlaceQualifier: Sendable {
    case on, `in`, near, at
}

/// A reverse-geocoded description of a point — nearest place name + qualifier.
/// Mirrors `domain.LocationDescription`.
public struct LocationDescription: Equatable, Sendable {
    public let title: String
    public let qualifier: PlaceQualifier?
    public let secondary: String?
    public let elevationM: Double?

    public init(title: String, qualifier: PlaceQualifier? = nil, secondary: String? = nil, elevationM: Double? = nil) {
        self.title = title
        self.qualifier = qualifier
        self.secondary = secondary
        self.elevationM = elevationM
    }

    /// Headline, e.g. "On Galdhøpiggen" / "In Lom".
    public var label: String {
        switch qualifier {
        case .on: "On \(title)"
        case .in: "In \(title)"
        case .near: "Near \(title)"
        case .at, nil: title
        }
    }

    /// Supporting line, e.g. "fjelltopp · 2469 m".
    public var subtitle: String {
        [secondary?.nonBlank, elevationM.map { "\(Int($0)) m" }]
            .compactMap { $0 }
            .joined(separator: " · ")
    }
}

private extension String {
    var nonBlank: String? { trimmingCharacters(in: .whitespaces).isEmpty ? nil : self }
}
